packages/ember-application/lib/utils/validate-type.js

```diff
@@ -0,0 +1,24 @@
+/**
+@module ember
+@submodule ember-application
+*/
+
+import Route from "ember-routing/system/route";
+import Component from "ember-views/views/component";
+import View from "ember-views/views/view";
+import Service from "ember-runtime/system/service";
+
+let VALIDATED_TYPES = {
+  route: Route,
+  component: Component,
+  view: View,
+  service: Service
+};
+
+export default function validateType(resolvedType, parsedName) {
+  let expectedType = VALIDATED_TYPES[parsedName.type];
+
+  Ember.assert(`Expected ${parsedName.fullName} to resolve to an ${expectedType} but instead it was ${resolvedType}.`, function() {
+    return !expectedType || expectedType.detect(resolvedType);
```

Comment:
I would prefer that we stamp the objects in question, instead of checking that they are instances.

The idea here is that at some point in the not terribly distant future, we do not want components/views to have to be a specific instance (they might be "POJO's" that adhere to our view systems requirements).  `Ember.Component` already adds `isComponent` to the instances, we could do the same to the factories.

What do you think?